### ISSUE TITLE
Varchar data truncation when fast_executemany is true

### ISSUE THREAD

**Issue by keitherskine** [mkleehammer/pyodbc#480]
### Environment

- Python: 3.6.5
- pyodbc: 4.0.24
- OS: Windows 10 64-bit
- DB: SQL Server Express (13.0.1742.0)
- driver: ODBC Driver 17 for SQL Server

### Issue

When `fast_executemany` is true, strings that should be too long to be inserted into the table are still inserted into the table, but truncated.

Here is some code that inserts increasingly long strings into a column that is just 5 characters max:

```python
cnxn = pyodbc.connect(MY_CNXN_STR, autocommit=True)
crsr = cnxn.cursor()

sql = "create table dbo.pyodbc_fastexecutemany (id int not null, str1 nvarchar(5) null)"
crsr.execute(sql)

sql = "insert into dbo.pyodbc_fastexecutemany (id, str1) values (?, ?)"

# ids from 1 to 10
data = [(i, 'a' * i) for i in range(1, 11)]
crsr.fast_executemany = False
try:
    crsr.executemany(sql, data)
except Exception as e:
    print(e)

# ids from 1001 to 1010
data = [(i+1000, 'a' * i) for i in range(1, 11)]
crsr.fast_executemany = True
try:
    crsr.executemany(sql, data)
except Exception as e:
    print(e)
```
Here's what ends up in the table:
```
id	str1
1	a
2	aa
3	aaa
4	aaaa
5	aaaaa
1001	a
1002	aa
1003	aaa
1004	aaaa
1005	aaaaa
1006	aaaaa
1007	aaaaa
1008	aaaaa
1009	aaaaa
1010	aaaaa
```
Note the length of the strings for ids 1006 to 1010.

I can kind of understand that when `fast_executemany` is false and `autocommit` is true, the first 5 records would be inserted into the table and the sixth would error out.  However, what seems very odd is that when `fast_executemany` is true, strings which are too long for the column are apparently truncated before inserting them anyway.  And this happens even though it appears some kind of error is returned by the database.  I would have thought the whole transaction should be rolled back, so that the end result is no records with ids 1000+ at all would end up in the table.  At the very least, string values should not be truncated.

Also, when `fast_executemany` is false, this error is printed out:
```
('22001', '[22001] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]String or binary data would be truncated. (8152) (SQLExecDirectW); [22001] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]The statement has been terminated. (3621)')
```
...and when `fast_executemany` is true, this error is printed out:
```
<built-in method executemany of pyodbc.Cursor object at 0x000001CB1CDF1738> returned a result with an error set
```

So there's two things here when `fast_executemany` is true:

1) Where are the strings being truncated?  By pyodbc, or at the server?  And is the string length being checked by pyodbc?
2) The error message returned in this case is not very helpful (although it has to be said there is more info if you use `traceback`).

**Comment by keitherskine** [comment]
Let me add a third question which is more of a general question about how `executemany` should behave when `fast_executemany` is true.  If someone is inserting a million records into a table, and just one field in one of those million records is too long by just one character, should the entire million inserts be rolled back?  Or should that one record not be inserted?  Or should that one field be truncated, and the record inserted anyway?  My inclination is to say this scenario should be treated as one database transaction and as such should be rolled back in its entirety.  It's difficult to see how a meaningful error message could be returned to the user otherwise (especially if it's not just one record that failed but hundreds of thousands of them that failed).

**Comment by keitherskine** [comment]
One further observation.  In the first part of the test above, when `fast_executemany` is false and `autocommit` is true, it does seem a bit odd to me that the first five records are inserted.  It would seem more logical for either all ten records to be inserted or none of the records be inserted.  That is, either the `executemany` call should completely succeed or completely fail.  What's worse, even when those five records are inserted, the cursor `rowcount` value returned is -1 so it's impossible to tell how many of the records were actually inserted.  In this case, it would be nice if the returned `rowcount` value was 5, not -1.

**Comment by keitherskine** [comment]
One last observation, if `executemany` is run with an UPDATE statement, the `rowcount` returned is always -1 (in the environment described above) whether `fast_executemany` is true or false.  This may best be raised as a separate issue, but is it possible to return the total number of records updated rather than -1?  I appreciate the `rowcount` value in this case might not be accurate because the same records might have been updated multiple times, but still, some kind of number is better than -1.

**Comment by v-chojas** [comment]
Please post an ODBC trace to allow better investigation into this issue.

> This may best be raised as a separate issue, but is it possible to return the total number of records updated rather than -1?

I believe that comes from the ODBC driver, not pyODBC.

**Comment by keitherskine** [comment]
Here's the trace for that initial code example.
[pyodbc_issue480_trace.txt](https://github.com/mkleehammer/pyodbc/files/2526475/pyodbc_issue480_trace.txt)

**Comment by v-chojas** [comment]
Please try the PR #482 code

**Comment by mkleehammer** [comment]
Merged into 4.0.25.  Thanks everyone.
